### ISSUE TITLE
Design criteria to test the performance and quality of found solutions

### ISSUE THREAD

**Issue by OpusV** [Mantella/Mantella#245]
Design criteria for comparison with other optimisation libraries.

**Comment by SebastianNiemann** [comment]
@OpusV I am currently working on this and created a first draft for the general process

``` bash
do_benchmark() {
  echo "${MAGENTA_TEXT_COLOR}Compiling and running benchmarks.${RESET_TEXT_COLOR}"
  
  cd ./benchmark || exit 1
  
  # List of library to benchmark
  # Loop through each library (replacing 'mantella' the library's name)
    cd ./mantella || exit 1
    
    if ! sudo docker build -t benchmark/mantella:latest .; then AN_ERROR_OCCURED=1; fi
    if ! sudo docker run -v "$(pwd)":/mantella -w /mantella --name benchmark_mantella -t -d benchmark/mantella; then AN_ERROR_OCCURED=1; fi
    
    sudo docker exec benchmark_mantella /bin/bash -c " \
      if [ ! -d "./build" ]; then mkdir build; fi && \
      cd ./build && \
      cmake .. && \
      make clean benchmark && \
      ./benchmark
    "
    
    # Post-processing needed?

    cd .. || exit 1
  # End loop
  
  # What do we want to do with the results?
  
  cd .. || exit 1
  
  finish_up
}
```

**Comment by SebastianNiemann** [comment]
Beside the aforementioned process, I think we should measure the needed computation time to reach an acceptable objective value as well as the best found objective value for a limited number of evaluations (not algorithm iterations, as the later on is implementation and not problem dependent - which might not be comparable at all).

Since most users will be interested on release-build performance, we should build all libraries with compile-optimisation (`-O3 -march=native -ffast-math`) and include optimised dependencies if needed (like MKL or OpenBLAS instead of the standard Netlib BLAS implementation).

**Comment by OpusV** [comment]
I thought about the same criteria. Yes, measuring the number of evaluation should be clearer instead of iterations if we'll have to possibility to abort the execution after reaching an acceptable value after a single evaluation. Compiler flags are fine whereas `-ffast-math` could be changed in another math flag to check if  the best found objective value is dependend of that... just a thought, but in general your ideads seem legit in my opinion.

**Comment by SebastianNiemann** [comment]
Yes, I have the same concerns about `-ffast-math`. In most projects (there are also alot of talks about it), there are no issues in using it (even more, the Intel C compiler activates it per default). However, some implementations might focus more on precision critical results then computation time.

**Comment by SebastianNiemann** [comment]
This is now implemented by #269
